src/lib/Libattr/attr_get_set.c

```diff
@@ -0,0 +1,290 @@
+ * "OpenPBS®", "PBS Professional®", and "PBS Pro™" and Altair's logos is
+ * subject to Altair's trademark licensing policies.
+ */
+
+
+#include <pbs_config.h>   /* the master config generated by configure */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "pbs_ifl.h"
+#include "attribute.h"
+#include "pbs_error.h"
+#include "libpbs.h"
+
+
+/**
+ * @brief	Generic attribute setter function, accepts all values as string regardless of the type
+ * 			Tip: use this when you want at_set() and at_decode() to be invoked, otherwise use the
+ * 			type based setters below
+ *
+ * @param[in]	pattr	-	pointer to attribute being set
+ * @param[in]	pdef 	-	attribute definition
+ * @param[in]	pobj	-	pointer to attribute definition
+ * @param[in]	value	-	value to be set
+ *
+ * @return	int
+ * @retval	0 for success
+ * @retval	1 for failure
+ *
+ * @par MT-Safe: No
+ * @par Side Effects: None
+ *
+ */
+int
+set_attr_generic(attribute *pattr, attribute_def *pdef, char *value, char *rescn, enum batch_op op)
+{
+	int rc;
+	attribute tempat;
+
+	if (pattr == NULL || pdef == NULL) {
+		log_err(-1, __func__, "Invalid pointer to attribute or its definition");
+		return 1;
+	}
+
+	/* Just call decode and set the value of attribute directly */
+	if (op == INTERNAL) {
+		if ((rc = pdef->at_decode(pattr, pdef->at_name, rescn, value)) != 0) {
+			log_errf(rc, __func__, "decode of %s failed", pdef->at_name);
+			return 1;
+		}
+		return 0;
+	}
+
+	clear_attr(&tempat, pdef);
+	if ((rc = pdef->at_decode(&tempat, pdef->at_name, NULL, value)) != 0) {
+		log_errf(rc, __func__, "decode of %s failed", pdef->at_name);
+		return 1;
+	}
+
+	rc = set_attr_with_attr(pdef, pattr, &tempat, op);
+
+	pdef->at_free(&tempat);
+
+	return rc;
+}
+
+/**
+ * @brief	Set attribute using another attribute
+ *
+ * @param[in]	pdef 	-	attribute definition
+ * @param[in]	oattr	-	pointer to attribute being set
+ * @param[in]	nattr	-	pointer to attribute to set with
+ * @param[in]	op		-	operation to do
+ *
+ * @return	int
+ * @retval	0 for success
+ * @retval	1 for failure
+ *
+ * @par MT-Safe: No
+ * @par Side Effects: None
+ *
+ */
+int
+set_attr_with_attr(attribute_def *pdef, attribute *oattr, attribute *nattr, enum batch_op op)
+{
+	int rc;
+
+	if ((rc = pdef->at_set(oattr, nattr, op)) != 0)
+		log_errf(rc, __func__, "set of %s failed", pdef->at_name);
+
+	return rc;
+}
+
+/**
+ * @brief	Attribute setter function for long type values
+ *
+ * @param[in]	pattr	-	pointer to attribute being set
+ * @param[in]	value	-	value to be set
+ *
+ * @return	void
+ *
+ * @par MT-Safe: No
+ * @par Side Effects: None
+ *
+ */
+void
+set_attr_l(attribute *pattr, long value, enum batch_op op)
+{
+	if (pattr == NULL) {
+		log_err(-1, __func__, "Invalid pointer to attribute");
+		return;
+	}
+
+	switch (op) {
+		case SET:
+			pattr->at_val.at_long = value;
+			break;
+		case INCR:
+			pattr->at_val.at_long += value;
+			break;
+		case DECR:
```

Comment:
why don't we also support unsetting the attribute using setter functions? you can just call at_free here and set  the flags appropriately.

Reply:
I took inspiration from our at_set() functions for this, they don't support an UNSET op. I took a quick look at mgr_unset_attr() but I didn't really see an obvious place for it. If you think that it might be useful, can you please point me to some code which it might help?

Reply:
set_attr_generic() just calls at_decode() and at_set() (both of which take 'op'), it doesn't have an explicit switch case for op. So I'd rather add a new function called unset_attr(), but if it's only going to be used in one place, mgr_unset_attr(), then I don't really see much value in it. I feel like mgr_unset_attr() already is a good handler for all attribute unset operations. What do you think?